qadence/ml_tools/callbacks/writer_registry.py

```
@@ -0,0 +1,375 @@
+            msg = f"Iteration {result.iteration: >7} | {loss_key.title()}: {loss_value:.7f} -"
+        else:
+            msg = f"Iteration {result.iteration: >7} | Loss: None -"
+        msg += " ".join([f"{k}: {v:.7f}" for k, v in result.metrics.items() if k != loss_key])
+        print(msg)
+
+
+class TensorBoardWriter(BaseWriter):
+    """Writer for logging to TensorBoard.
+
+    Attributes:
+        writer (SummaryWriter): The TensorBoard SummaryWriter instance.
+    """
+
+    def __init__(self) -> None:
+        self.writer = None
+
+    def open(self, config: TrainConfig, iteration: int = None) -> SummaryWriter:
+        """
+        Opens the TensorBoard writer.
+
+        Args:
+            config: Configuration object containing settings for logging.
+            iteration (int, optional): The iteration step to start logging from.
+                Defaults to None.
+
+        Returns:
+            SummaryWriter: The initialized TensorBoard writer.
+        """
+        log_dir = str(config._log_folder)
+        if isinstance(iteration, int):
+            self.writer = SummaryWriter(log_dir=log_dir, purge_step=iteration)
+        else:
+            self.writer = SummaryWriter(log_dir=log_dir)
+        return self.writer
+
+    def close(self) -> None:
+        """Closes the TensorBoard writer."""
+        if self.writer:
+            self.writer.close()
+
+    def write(self, result: OptimizeResult) -> None:
+        """
+        Logs the results of the current iteration to TensorBoard.
+
+        Args:
+            result (OptimizeResult): The optimization results to log.
+        """
+        # Not writing loss as loss is available in the metrics
+        # if result.loss is not None:
+        #     self.writer.add_scalar("loss", float(result.loss), result.iteration)
+        if self.writer:
+            for key, value in result.metrics.items():
+                self.writer.add_scalar(key, value, result.iteration)
```

what happens in the non-happy path? (`bool(self.writer) is False`)? Should we include an error?

---

Sure. We can raise a runtime error.